WordPress/gutenberg
# Experiment: Auto-inserting blocks on the frontend

## What?
Early-stages experiment to explore concepts discussed in #39439.

## Why?
See #39439. In short, block themes are currently lacking extensibility; the concept of hooks and filters from PHP themes doesn't carry over to them, so we need a more block-centric concept instead.

## How?
Per discussion in #39349, we'd likely want blocks to be auto-inserted as the (previous or next) sibling of a given block, or as its (first or last) child. In this PR, we're demonstrating one example of each sibling and child insertion:
- We use the `render_block` hook to insert the Social Icons block as the next sibling of the Post Content block.
- We use the `render_block_data` hook (which -- unlike the `render_block` hook -- conveniently gives us access to a given block's children) to insert the Avatar block as the last child of the Comment Template block.

The latter example is chosen to also demonstrate that block context is successfully passed to auto-inserted blocks (as evidenced by the auto-inserted Avatar blocks showing each comment author's avatar).

This PR required some preparation in order to make the Comment Template block work with auto-inserted blocks, see in particular https://github.com/WordPress/gutenberg/pull/50279, https://github.com/WordPress/gutenberg/pull/50879, and https://github.com/WordPress/gutenberg/pull/50883. (Similar modifications were applied to the Post Template block, see https://github.com/WordPress/gutenberg/pull/50313.)

## Testing Instructions

- Use the Twenty Twenty-Three theme.
- On a single post page, you should see the Social Icon block (showing the WordPress logo) below the post (as sort of a stand-in for a Like button 😅)
- Have a look at comments below a given post. You should see an Avatar block showing the comment author's avatar (as a stand-in for a Comment Like button)

## Screenshots or screencast

![image](https://github.com/WordPress/gutenberg/assets/96308/466860cd-f3f7-4168-9b47-a4414951dd6d)
